pkg/ipcache/ipcache.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipcache
+
+import (
+	"encoding/json"
+	"path"
+	"sort"
+	"sync"
+
+	"github.com/cilium/cilium/pkg/envoy"
+	envoyAPI "github.com/cilium/cilium/pkg/envoy/cilium"
+	"github.com/cilium/cilium/pkg/envoy/xds"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// DefaultAddressSpace is the address space used if none is provided.
+	DefaultAddressSpace = "default"
+)
+
+var (
+	// IPIdentitiesPath is the path to where endpoint IPs are stored in the key-value
+	//store.
+	IPIdentitiesPath = path.Join(kvstore.BaseKeyPrefix, "state", "endpointIPs", "v1")
+
+	// IPIdentityCache caches the mapping of endpoint IPs to their corresponding
+	// security identities across the entire cluster in which this instance of
+	// Cilium is running.
+	IPIdentityCache = NewIPCache(envoy.NetworkPolicyHostsCache)
+
+	// AddressSpace is the address space (cluster, etc.) in which policy is
+	// computed. It is determined by the orchestration system / runtime.
+	AddressSpace = DefaultAddressSpace
+
+	setupIPIdentityWatcher sync.Once
+)
+
+// IPCache is a caching of endpoint IP to security identity (and vice-versa) for
+// all endpoints which are part of the same cluster.
+type IPCache struct {
+	Mutex              lock.RWMutex
```

Comment:
Do we really need this mutex to be public?

Reply:
Nope - fixed.